### ISSUE TITLE
Travis-CI Node version usage stats.

### ISSUE THREAD

**Issue by jdalton** [travis-ci/travis-ci#6659]
I was wondering if the Travis-CI team had insights into its Node 0.10 usage?
Usage stats help developers gauge when to drop support for older Node versions.

👉 It's good to know the extent by which Node version usage stats are distorted by CIs.
#### Context:

From time to time @seldo, of npm, [posts npm usage stats](https://twitter.com/seldo/status/781258069678391296) for various Node versions.
In a recent update the usage stats for Node 0.10, which [falls off Node support](https://github.com/nodejs/LTS#lts-schedule) in October,
were **~13%**.

> <img width="60%" src="https://cloud.githubusercontent.com/assets/4303/19008647/770015ba-8721-11e6-8aa1-9f3b62996431.png" title="click to enlarge">

Drilling in a bit more, it appears **90%** of those on 0.10 come from Linux _(maybe from Travis?)_.
\@thealphanerd 

> <img width="50%" src="https://cloud.githubusercontent.com/assets/4303/19008711/063d85a0-8722-11e6-9c07-273c52fe98b8.png" title="click to enlarge">

**Comment by jdalton** [comment]
Using the [Google BigQuery Github dataset](https://cloud.google.com/bigquery/public-data/github)
I created a query to rank the node versions specified in `.travis.yml` files:
```sql
SELECT 
  COUNT(*) as cnt, version
FROM 
  JS(
    (SELECT content FROM [bigquery-public-data:github_repos.contents] WHERE id IN (
      SELECT id FROM [bigquery-public-data:github_repos.files] WHERE RIGHT(path, 12) = ".travis.yml" 
    )),
    content,
    "[{ name: 'version', type: 'string'}]",
    "function(row, emit) {
      try {
        var yml = row.content;
        var match = yml.match(/^ *node_js\s*:((?:\s*-.+)+|\s*\[[\s\S]+?\])/m);
        var snippet = match ? match.pop() : '';
        var versions = snippet.match(/[\d.]+/g);

        versions.forEach(function(ver) {
          var major = ver.match(/^(?:0\.)*\d+/)[0];
          emit({ 'version': major });
        });
      } catch (e) {}
    }" 
  )
GROUP BY version
ORDER BY cnt DESC
LIMIT 1000
```

The results show a large number of 0.10 runs:
<img src="https://cloud.githubusercontent.com/assets/4303/19009544/e8474d44-8729-11e6-9b87-853ec9a83a38.png">
Which is consistent with the npm usage stat. @seldo is there a way to filter out Travis and other CIs IP ranges from the npm usage data?

**Comment by Joshua-Anderson** [comment]
Travis CI runs on aws and gcs, so I think filtering IPs is probably impossible to do without excluding all the production applications that run on these platforms.

**Comment by jdalton** [comment]
@Joshua-Anderson

Could you double check.

Via @seldo:

[status/782037773150629888](https://twitter.com/seldo/status/782037773150629888):
> Their IPs are extremely easy to detect if you are the npm registry :-)

[status/782038480993996800](https://twitter.com/seldo/status/782038480993996800):
> They have 4 main IPs, linked to http://travis-ci.com  hostnames, that account for 10% of traffic.

**Comment by Joshua-Anderson** [comment]
@jdalton You're right, I'm a little out of date: https://docs.travis-ci.com/user/ip-addresses/ Travis-CI doesn't  give an explicit range for GCS yet though.

**Comment by jdalton** [comment]
@Joshua-Anderson

To summarize [docs.travis-ci.com/user/ip-addresses](https://docs.travis-ci.com/user/ip-addresses/), there's potentially 5+ IPs?

>
As of mid March 2016:
>
Container-based (travis-ci.com):
  1) 54.172.141.90/32
  2) 52.3.133.20/32
>
Container-based (travis-ci.org):
  3) 52.0.240.122/32
  4) 52.22.60.255/32
>
OS X:
  5) 208.78.110.192/27

👉 The unknown is IPs for sudo-enabled Linux.

I performed a quick query to see how many are sudo enabled.
There's a gotcha though. The docs state that the default for unspecified matches [changed](https://docs.travis-ci.com/user/ci-environment/#Virtualization-environments) some time in 2015 so this query doesn't account for that if that's the case.

##### SQL query:
<details>
```sql
SELECT 
  COUNT(*) as cnt, sudo
FROM 
  JS(
    (SELECT content FROM [bigquery-public-data:github_repos.contents] WHERE id IN (
      SELECT id FROM [bigquery-public-data:github_repos.files] WHERE RIGHT(path, 11) = ".travis.yml" 
    )),
    content,
    "[{ name: 'sudo', type: 'boolean'}]",
    "function(row, emit) {
       var yml = row.content || '';
       var isNode = /^\s*node_js\s*:(?:(?:\s*-.+)+|\s*\[[\s\S]+?\])/m.test(yml);
       
       if (isNode) {
         var match = yml.match(/^\s*sudo\s*:\s*(.+)/m);
         var snippet = match ? match[1] : '';
         var cleaned = snippet.replace(/#.+/g, '').trim();

         var sudo = cleaned != 'false' && cleaned != 'no';
         emit({ sudo: sudo });
       }
     }" 
  )
GROUP BY sudo
ORDER BY cnt DESC
```
</details>

The results show there are considerably more sudo runs.
In other words more IPs which aren't available than those that are.
>
<img width="139" src="https://cloud.githubusercontent.com/assets/4303/19014892/de996730-87ad-11e6-9691-e62b55a82015.png">

Does that gel with Travis' data?

**Comment by jdalton** [comment]
Focusing in on just 0.10 it looks like more than 4x of them are using sudo.

##### SQL query:
<details>
```sql
SELECT 
  COUNT(*) as cnt, sudo
FROM 
  JS(
    (SELECT content FROM [bigquery-public-data:github_repos.contents] WHERE id IN (
      SELECT id FROM [bigquery-public-data:github_repos.files] WHERE RIGHT(path, 11) = ".travis.yml" 
    )),
    content,
    "[{ name: 'sudo', type: 'boolean'}]",
    "function(row, emit) {
       var yml = row.content || '';
       var match = yml.match(/^\s*node_js\s*:((?:\s*-.+)+|\s*\[[\s\S]+?\])/m);
       var snippet = match ? match[1] : '';
       var cleaned = snippet.replace(/#.+/g, '');
       var versions = cleaned.match(/[\w.&*]+/g) || [];

       versions = versions.map(function(ver) {
         var match = ver.match(/^v?([0.]*\d+)/);
         var snippet = match ? match[1] : '';
         var cleaned = snippet.replace(/\b0{2,}\b/g, '0').replace(/^\./, '0.');
         var major = cleaned || ver.replace(/\bio\.js\b/g, 'iojs');
         return major;
       });

       if (versions.indexOf('0.10') < 0) {
         return;
       }
       var match = yml.match(/^\s*sudo\s*:\s*(.+)/m);
       var snippet = match ? match[1] : '';
       var cleaned = snippet.replace(/#.+/g, '').trim();

       var sudo = cleaned != 'false' && cleaned != 'no';
       emit({ sudo: sudo });
     }" 
  )
GROUP BY sudo
ORDER BY cnt DESC
```
</details>

>
<img width="137" src="https://cloud.githubusercontent.com/assets/4303/19016054/11c52dc6-87c5-11e6-9c3d-2288042a2972.png">

**Comment by gr2m** [comment]
This is great work, thanks for doing this!
